### PR TITLE
Issue #2779: DATE_PART structs

### PR DESCRIPTION
**Pull request by hawkfish**
Implement for TIMESTAMP, DATE, TIME, INTERVAL and TIMESTAMPTZ.
Tweak promotion weights to prefer VARCHAR to LIST.
Some minor code cleanup and templatisation.
Expand test data.

### CODE REVIEW

File: test/sql/function/timestamp/test_icu_datepart.test [hunk 1]
```diff
@@ -284,13 +325,61 @@ SELECT weekday(ts), weekday(ts::TIMESTAMP) FROM timestamps;
...
+{'era': 1, 'year': 2020, 'month': 12, 'day': 31}	2021-01-01 00:00:00+00
+{'era': 1, 'year': 2021, 'month': 2, 'day': 1}	2021-02-02 00:00:00+00
+{'era': 1, 'year': 2021, 'month': 11, 'day': 15}	2021-11-15 09:30:00+00
+{'era': 1, 'year': 2021, 'month': 11, 'day': 15}	2021-11-15 10:30:00+00
+{'era': 1, 'year': 2021, 'month': 11, 'day': 26}	2021-11-26 10:15:13.123456+00
+{'era': 1, 'year': 2021, 'month': 12, 'day': 24}	2021-12-24 22:00:00+00
+
+# Time parts
+query II
+SELECT DATE_PART(['hour', 'minute', 'microsecond', 'offset'], ts), ts
```
Comment (Mytherin) [inline test/sql/function/timestamp/test_icu_datepart.test:358]: Could you add tests for the following scenarios:

* What if the list of parts is empty?
* What about nulls in the list of parts?
* What about non-constants int he list of parts?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good. A request for more tests below. Also, since the purpose of this is to improve performance when multiple extractions are requested, could we add some benchmarks comparing multiple calls to date_part to a single call to date_part with multiple components? Particularly I'm interested in the cross-over point (is it already faster with two parts, or three parts, etc), since there does seem to be some extra work performed per value.

**Comment by hawkfish** [comment]
Filling in a `STRUCT` is a row-ish operation, so there is definitely a trade off. I was expecting that the multiplexing was less costly than all the arithmetic, but I should measure it.

**Comment by hawkfish** [comment]
I see parity at about 5 (['year', 'month', 'day', 'decade', 'week']), but they both seem to stop there. Profiling is indicated.

It's also worth mentioning that ICU computes all parts when you ask for any of them, but the built in code only computes a few. Is it possible to benchmark ICU calls?

**Comment by Mytherin** [comment]
You can add `require icu` to a benchmark script, similar to a test script.

**Comment by hawkfish** [comment]
The crossover depends on what calculations are being performed. In particular, the week calculations are expensive and should only be performed if needed. After filtering computations for the struct, I see the crossover for the YMD family at around 4 values (year, month, day, decade are on par, but adding century puts struct ahead:
[Crossover Dashboard](https://public.tableau.com/views/DatePartStructCrossover/CrossoverDashboard?:language=en-GB&publish=yes&:display_count=n&:origin=viz_share_link)

**Comment by Mytherin** [comment]
Perhaps it could be sped up by having a `bool part_required[X]` array that is used to figure out which values are actually required needed? I imagine certain components are also much more common than others (e.g. I imagine decade, century and millennium are very rarely used).

**Comment by Mytherin** [comment]
Thanks for the updates! This looks great now. One minor nit: there are several combinations of date parts that appear to not be tested, e.g. the blocks with `mask & DOW`, `mask & EPOCH` and `mask & DOY`. Could you add some tests that trigger these scenarios? After that I think this is ready to merge.

**Comment by hawkfish** [comment]
Good call - found three small bugs!

**Comment by Mytherin** [comment]
Excellent! Thanks for the updates. Everything looks now.

**Pull request merged**